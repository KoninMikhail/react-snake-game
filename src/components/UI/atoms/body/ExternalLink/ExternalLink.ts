import styled from 'styled-components';

/**
 * @typedef IExternalLinkProps
 *
 * @property {string} href - The href of the external link
 * @property {string=} - target - The target of the external link
 * @property {string=} - rel - The rel of the external link
 * @property {JSX.Element=} - children - The children of the external link.
 */
interface IExternalLinkProps {
  href: string;
  target?: string;
  children?: JSX.Element | JSX.Element[];
  rel?: string;
}

/**
 * ExternalLink
 *
 * @description Use for external links with <a> tags
 *
 * @param {IExternalLinkProps} props - Properties
 * @returns {JSX.Element} External link element.
 */

export const ExternalLink = styled.a.attrs<IExternalLinkProps>((props) => ({
  href: props.href || '#',
  target: props.target || '_blank',
  rel: props.rel || 'noopener noreferrer',
}))`
  color: white;
  text-decoration: none;
  transition: all 0.3s ease-out;

  &:hover {
    text-decoration: underline;
  }
`;
