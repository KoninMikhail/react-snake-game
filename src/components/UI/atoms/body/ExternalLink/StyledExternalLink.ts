import styled from 'styled-components';
import { IExternalLinkProps } from './ExternalLink';

export const StyledExternalLink = styled.a.attrs((props) => ({
  href: props.href || '#',
  target: props.target || '_blank',
  rel: props.rel || 'noopener noreferrer',
  children: props.children || 'link',
}))<IExternalLinkProps>`
  cursor: pointer;
  color: white;
  text-decoration: none;
  transition: all 0.3s ease-out;

  &:hover {
    text-decoration: underline;
  }
`;
