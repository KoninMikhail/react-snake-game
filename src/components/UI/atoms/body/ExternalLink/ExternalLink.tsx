import { StyledExternalLink } from './StyledExternalLink';

/**
 * @typedef IExternalLinkProps
 *
 * @property {string} href - The href of the external link
 * @property {string=} - target - The target of the external link
 * @property {string=} - rel - The rel of the external link
 * @property {JSX.Element=} - children - The children of the external link.
 */
export interface IExternalLinkProps {
  href?: string;
  target?: string;
  children?: JSX.Element | JSX.Element[] | string;
  rel?: string;
}

/**
 * ExternalLink
 *
 * Use for external links with <a> tags
 *
 * @param {IExternalLinkProps} props - Properties
 * @returns {JSX.Element} External link element.
 */

export const ExternalLink = ({ children, ...rest }: IExternalLinkProps): JSX.Element => (
  <StyledExternalLink {...rest}>{children}</StyledExternalLink>
);
