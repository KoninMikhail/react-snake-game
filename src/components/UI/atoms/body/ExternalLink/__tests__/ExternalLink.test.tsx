import { render, screen } from '@testing-library/react';
import { ExternalLink } from '../ExternalLink';

describe('ExternalLink', () => {
  it('correctly render with props', () => {
    render(
      <ExternalLink href="https://www.google.com/" target="_self" rel="nofollow">
        {'yeah!'}
      </ExternalLink>
    );

    expect(screen.getByText('yeah!')).toBeInTheDocument();
    expect(screen.getByText('yeah!')).toHaveAttribute('href', 'https://www.google.com/');
    expect(screen.getByText('yeah!')).toHaveAttribute('target', '_self');
    expect(screen.getByText('yeah!')).toHaveAttribute('rel', 'nofollow');
  });
  it('correctly render without props', () => {
    render(<ExternalLink />);

    expect(screen.getByText(/link/i)).toBeInTheDocument();
    expect(screen.getByText(/link/i)).toHaveAttribute('href', '#');
    expect(screen.getByText(/link/i)).toHaveAttribute('rel', 'noopener noreferrer');
    expect(screen.getByText(/link/i)).toHaveAttribute('target', '_blank');
  });

  it('equal snapshot', () => {
    const { container } = render(
      <ExternalLink href="https://www.google.com/" target="_self" rel="nofollow">
        {'yeah!'}
      </ExternalLink>
    );

    expect(container).toMatchSnapshot();
  });
});
