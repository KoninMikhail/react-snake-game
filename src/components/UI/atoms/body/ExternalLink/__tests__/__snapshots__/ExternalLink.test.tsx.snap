// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ExternalLink equal snapshot 1`] = `
.c0 {
  cursor: pointer;
  color: white;
  -webkit-text-decoration: none;
  text-decoration: none;
  -webkit-transition: all 0.3s ease-out;
  transition: all 0.3s ease-out;
}

.c0:hover {
  -webkit-text-decoration: underline;
  text-decoration: underline;
}

<div>
  <a
    class="c0"
    href="https://www.google.com/"
    rel="nofollow"
    target="_self"
  >
    yeah!
  </a>
</div>
`;
